### PR TITLE
Tandem Feature Cross: config-aware sigmoid gate on encoded features

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Gap and stagger scalars currently feed only into `--gap_stagger_spatial_bias` (slice routing). This helps routing but doesn't affect the **INPUT feature encoding**. For OOD NACA6416: the DSDF camber features signal a different geometric prior, but the model has no mechanism to upweight these features when it detects an OOD configuration.

A **sigmoid gate parameterized by (gap, stagger, Re, AoA)** can learn "for this tandem gap/stagger, amplify/suppress specific encoded feature channels." The gate is a global vector (same for all N nodes in a sample) conditioned on 4 flow configuration scalars. Near-identity initialization (`sigmoid(5) ≈ 0.99`) ensures first pass is baseline-equivalent.

This is distinct from the failed `feature_cross` from earlier phases (PR #1436) which was a learned node-to-node interaction matrix. This is a **GLOBAL gate** (same vector broadcast over all N nodes) conditioned on 4 scalars — much lower risk of overfitting.

**Physical motivation:** Different tandem configurations create qualitatively different flow physics. At small gaps, wake impingement dominates; at large gaps, the foils are nearly independent. The model should modulate which encoded features are emphasized based on the configuration — this gate provides that mechanism.

**Confidence:** Medium. The mechanism is physically motivated. Gap and stagger as input-gate conditioning is novel in this codebase (previously only used in routing). Near-identity init is safe. Main risk: redundancy with `--gap_stagger_spatial_bias` (routing vs feature modulation are mechanistically different, so orthogonality is plausible).

## Instructions

All changes in `cfd_tandemfoil/train.py`.

### Step 1: Add config flag

In the `Config` dataclass:
```python
tandem_feature_cross: bool = False   # config-aware sigmoid gate on encoded features
```

### Step 2: Create the gate module

In the model initialization section (near where `gap_stagger_spatial_bias` components are created), add:

```python
tandem_gate = None
if cfg.tandem_feature_cross:
    tandem_gate = nn.Sequential(
        nn.Linear(4, 32),
        nn.GELU(),
        nn.Linear(32, cfg.n_hidden),
    ).to(device)
    # Near-identity init: sigmoid(5) ≈ 0.993 → gate ≈ 1 (near-passthrough)
    nn.init.zeros_(tandem_gate[-1].weight)
    nn.init.constant_(tandem_gate[-1].bias, 5.0)
    tandem_gate = torch.compile(tandem_gate, mode=cfg.compile_mode)
    _gate_n = sum(p.numel() for p in tandem_gate.parameters())
    print(f"Tandem feature cross gate: {_gate_n:,} params (sigmoid(5) init)")
```

### Step 3: Add to optimizer

Add `tandem_gate` parameters to the optimizer. Match the LR of the main model:
```python
if tandem_gate is not None:
    param_groups[0]['params'].extend(list(tandem_gate.parameters()))
```

### Step 4: EMA tracking

Declare `ema_tandem_gate = None` alongside other EMA modules. In the EMA update loop, mirror the tandem_gate → ema_tandem_gate tracking.

### Step 5: Apply gate in training forward pass

After the input encoding (`preprocess` or equivalent), before the TransolverBlocks:

```python
if tandem_gate is not None:
    # Extract config scalars — these should be available from the data
    # gap: inter-foil gap, stagger: vertical offset, re: Reynolds number, aoa: angle of attack
    # You'll need to find where these scalars are available in the data pipeline
    # They may be in the raw input features or computed from the mesh geometry
    config_vec = torch.stack([gap, stagger, re_scalar, aoa_scalar], dim=-1)  # [B, 4]
    gate = torch.sigmoid(tandem_gate(config_vec))  # [B, n_hidden]
    fx = fx * gate[:, None, :]  # [B, N, n_hidden] element-wise
```

**Important:** For non-tandem samples (gap=0, stagger=0), the gate should be near-identity (sigmoid(bias=5)≈0.993). Verify this is the case.

**Finding the 4 config scalars:** Search the training loop for where gap, stagger, Re, AoA are available. They may be stored as part of the sample metadata or computed from `x`. Look for variables like `_gap`, `_stagger`, `_re`, `_aoa` or similar. If Re and AoA are not readily available as scalars, use 0.0 as placeholder and just condition on (gap, stagger, 0, 0) — even 2 scalars provide useful configuration context.

### Step 6: Apply gate in validation/eval loops

Mirror the gate application using `ema_tandem_gate` in all eval loops.

### Step 7: Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py \
  --agent nezuko --wandb_name "nezuko/tandem-feat-cross-s42" \
  --wandb_group "round17/tandem-feature-cross" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --tandem_feature_cross

# Seed 73 — identical but --seed 73 --wandb_name "nezuko/tandem-feat-cross-s73"
```

### Step 8: Report results

Add a **Results** comment to this PR with:
- Table: p_in, p_oodc, p_tan, p_re for seed 42, seed 73, and 2-seed average
- Comparison against baseline
- W&B run IDs
- What gate values the model converged to (log gate statistics if possible)

## Baseline

Current best (PR #2213, Wake Deficit Feature, 2-seed average):

| Metric | Baseline | Target to beat |
|--------|----------|----------------|
| p_in   | **11.979** | < 11.98 |
| p_oodc | **7.643**  | < 7.65  |
| **p_tan** | **28.341** | **< 28.34** |
| p_re   | **6.300**  | < 6.30  |

W&B runs: `hgml7i2r` (seed 42), `qic03vrg` (seed 73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent nezuko --wandb_name "nezuko/baseline-wake-deficit" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
## Experiment Results

### Setup
- **Hypothesis**: A config-aware sigmoid gate conditioned on (gap, stagger, log_Re, AoA) applied to the model's input features (dim=66) should let the model learn different feature weightings per operating condition, enabling better generalization across the tandem configuration space.
- **Implementation**: 4→32→66 MLP producing a per-channel sigmoid gate; near-identity init (bias=5.0 → sigmoid(5)≈0.993); applied to `x` after Fourier PE, before model call.
- **Runs**: 2 seeds (42 and 73), WANDB_MODE=offline

### Metrics

| Split | Seed 42 (zb4kh079) | Seed 73 (j70lmc3s) | 2-seed avg | Baseline (PR #2213) | Δ |
|---|---|---|---|---|---|
| p_in | **11.628** | 12.285 | **11.956** | 11.979 | **-0.023 (-0.2%)** ✓ |
| p_tan | 28.8 | 28.6 | **28.70** | 28.341 | **+0.36 (+1.3%)** ✗ |
| p_oodc | 7.5 | 7.6 | **7.55** | 7.643 | **-0.09 (-1.2%)** ✓ |
| p_re | 6.4 | 6.4 | **6.40** | 6.300 | **+0.10 (+1.6%)** ✗ |

Best epoch: s42=146, s73=147.

### Analysis

Mixed results. The gate yields tiny improvements on p_in (-0.2%) and p_oodc (-1.2%), but causes regressions on p_tan (+1.3%) and p_re (+1.6%). The 2-seed average does not beat baseline on the critical tandem-transfer and OOD-Re metrics.

The likely explanation: the gate is conditioned on raw config scalars (gap@x[0,22], stagger@x[0,23], log_Re@x[0,13], AoA@x[0,14]), which are available in training but whose gate weights may not generalize to unseen tandem/Re configurations (tandem transfer dataset has different gap/stagger distributions than in-distribution). At near-identity init, the gate should start neutral — yet it still regresses on generalisation splits, suggesting the gate learns to partially suppress features in a way that hurts OOD.

The approach of gating the full input feature vector before the model is probably too blunt: it modulates all 66 dimensions uniformly based on 4 scalars that are highly collinear with the normalisation statistics in training.

### Suggested follow-ups
1. **Gate only the geometry channels** (DSDF / SDF dimensions, not velocity/pressure features) — the gap and stagger scalars are geometry-context and should mostly modulate geometry encoding, not physics channels.
2. **Apply gate at a later layer** (e.g. inside TransolverBlock after attention, conditioned on the hidden state mean) — this is more of a FiLM-style modulation and avoids suppressing early input features.
3. **Deeper gate with cross-attention** between a config embedding and the node features — this is architecturally heavier but more expressive.
4. **Drop the gate altogether** and instead inject the 4 config scalars as additional AdaLN conditioning (already partly done via domain_layernorm) — the gate may be redundant with existing config conditioning.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2226 because ADVISOR: Closing. p_in -0.2% and p_oodc -1.2% are encouraging, but p_tan +1.3% is a deal-breaker — p_tan is our most critical metric and it clearly regressed.

Your analysis is spot on: the gate is too blunt. Conditioning all 66 input channels on 4 global scalars that are collinear with normalization statistics means the gate learns to optimize for the training distribution but hurts OOD generalization (p_tan, p_re both worse).

The p_oodc improvement is interesting — it suggests config-awareness CAN help for OOD conditions. But the mechanism needs to be more targeted (e.g., gate only geometry channels, or apply FiLM modulation deeper in the network).

Given our overall findings, I'm not sending this back for iteration. The config-aware gating direction has structural limitations (any global gate will struggle with OOD configs it hasn't seen). Moving you to a different approach..
